### PR TITLE
EH6-cards-and-server

### PR DESCRIPTION
**Pull request by LeStolz**
- Add action “drop” using the F key
- Rename action “ability_1” to “card_0” and “ability_2” to “card_1”
- Add Card class and Cards
- Add Empty Node scene for better control of entities
- Add Flip Position2D into Player to flip stuff
- Add remote func to queue_free homing missile
- Make homing missile synchronize via its own position instead of mouse position
- Rename some other stuff
- Rename “CaptureTheFlag” to “captureTheFlag”
- Change default_gravity to 3000
- Set collision_layer of CardEffects and Cards to bit 8
- Set collision_layer and mask of Flag, Explosion, Background and Foreground to their appropriate ones
- Change stuff in Player
- Restructure ServerHandler
- Add MathUtil
- Organize files

### CODE REVIEW

File: src/utils/MathUtil.gd [hunk 1]
```diff
@@ -0,0 +1,35 @@
+extends Object
```
Comment (Legacy107) [inline src/utils/MathUtil.gd]: Extends Reference for performance issue.

File: src/utils/MathUtil.gd [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+
+func calculate_pivot_rotation(Pivot, Target, Origin):
```
Comment (Legacy107) [inline src/utils/MathUtil.gd]: Make this a static func.

File: src/scripts/singletons/ServerHandler.gd [hunk 3]
```diff
@@ -38,58 +43,101 @@ func _connection_failed():
...
+	NodeInstance.set_name(str(player_id))
 
-	PlayerContainer.set_network_master(player_id)
-	PlayerContainer.set_name(str(player_id))
-	Player.set_network_master(player_id)
-	Player.position = WorldNode.get_node("Spawns/Spawn" + str(player_id % 4 + 1)).position
+	WorldNode.get_node("EntityContainer").add_child(NodeInstance)
 
-	WorldNode.get_node("Players").add_child(PlayerContainer)
+	var PlayerInstance = Player.instance()
```
Comment (Legacy107) [inline src/scripts/singletons/ServerHandler.gd]: Variable declaration should be at the top of the scope

File: src/scripts/cards/Card.gd [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+var card_cooldown
+var card_swivel_dampening
+
+
+onready var card_state = {
+	"card_effect_instance_id" : 0
+}
+
+
+func _init(card_info):
```
Comment (Legacy107) [inline src/scripts/cards/Card.gd]: Rename card_info and variable to card_config and setting.

File: src/scripts/others/Player.gd [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	"facing_right" : facing_right,
+	"animation" : "Idle",
+	"mouse_position" : get_global_mouse_position(),
+}
+
+
+func _ready():
+	ready = true
+
+	for _id in range (cards_size):
```
Comment (Legacy107) [inline src/scripts/others/Player.gd:51]: Use array.resize() to fill cards with Null

File: src/scripts/others/Player.gd [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+onready var Hand = $Flip/Hand
+onready var Mouse = $Mouse
+onready var Camera = $Camera2D
+onready var AnimationPlayer = $AnimationPlayer
+onready var WallClimbDebounce = $WallClimbDebounce
+
+
+var ready = false
+var team_id = 0
+var cards_size = 2
```
Comment (Legacy107) [inline src/scripts/others/Player.gd:30]: Is this a constant

File: src/scripts/others/Player.gd [hunk 7]
```diff
@@ -0,0 +1,179 @@
+class_name Player
+
+extends KinematicBody2D
+
+
+export var MOVE_SPEED = 400
+export var MAX_MOVE_SPEED = 600
+export var JUMP_FORCE = 800
+export var MAX_FALL_SPEED = 800
+export var GRAVITY = 3000
```
Comment (Legacy107) [inline src/scripts/others/Player.gd]: Set GRAVITY to ProjectSettings.get_setting("physics/2d/default_gravity")

File: src/scripts/others/Player.gd [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+export var FRICTION = 100
+
+
+onready var NodeUtil = preload("res://src/utils/NodeUtil.gd").new()
+
+
+onready var Container = get_node("/root/Game/World/EntityContainer/" + name)
+
+
+onready var Flip = $Flip
```
Comment (Legacy107) [inline src/scripts/others/Player.gd]: Rename to Visual

File: src/scenes/Game.tscn [hunk 9]
```diff
@@ -1,7 +1,7 @@
 [gd_scene load_steps=3 format=2]
 
 [ext_resource path="res://src/scenes/World.tscn" type="PackedScene" id=1]
-[ext_resource path="res://src/scripts/GameHandler.gd" type="Script" id=2]
+[ext_resource path="res://src/scripts/gamemodes/GameHandler.gd" type="Script" id=2]
```
Comment (Legacy107) [inline src/scenes/Game.tscn]: Revert this change

File: src/utils/MathUtil.gd [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+	# This func calculates the Pivot rotation such that Origin's look vector points at
+	# Target's global_position
+
+	# Pivot is the pivot and the object whose rotation is calculated in this func
```
Comment (Legacy107) [inline src/utils/MathUtil.gd]: What does "the pivot and the object" mean?

File: src/scripts/cardEffects/HomingMissile.gd [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	var desired = global_position.direction_to(get_global_mouse_position()) * MAX_SPEED
+	var force = velocity.direction_to(desired).clamped(MAX_FORCE)
+
+	apply_force(force)
+	velocity = (velocity + acceleration).clamped(MAX_SPEED)
+	position += velocity * delta
+
+	homing_missile_state["position"] = position
+
+	ServerHandler.synchronize_unreliable(get_path(),
```
Comment (Legacy107) [inline src/scripts/cardEffects/HomingMissile.gd]: Break lines like this instead
```
ServerHandler.synchronize_unreliable(
    get_path(),
    "update_homing_missile_state",
    homing_missile_state
)
```

File: src/scripts/cardEffects/HomingMissile.gd [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+var MAX_SPEED = 150
+var MAX_FORCE = 100
+
+
+var acceleration = Vector2()
+var velocity = Vector2()
+
+
+onready var homing_missile_state = {
+	"position" : position
```
Comment (Legacy107) [inline src/scripts/cardEffects/HomingMissile.gd]: Add trailing comma

File: src/scripts/cards/Card.gd [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+var card_aims_at_mouse
+var card_only_activate_on_click
+var card_can_be_dropped
+var card_charges
+var card_cooldown
+var card_swivel_dampening
+
+
+onready var card_state = {
+	"card_effect_instance_id" : 0
```
Comment (Legacy107) [inline src/scripts/cards/Card.gd]: Add trailing comma

File: src/scripts/cards/Card.gd [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+		handle_drop_card_event(event)
+
+	handle_equip_card_event(event, 0)
+	handle_equip_card_event(event, 1)
+
+
+func aim_at_mouse():
+	Tween.stop(Pivot, "rotation")
+
+	Tween.interpolate_property(
```
Comment (Legacy107) [inline src/scripts/cards/Card.gd:90]: Line break like this
```
Tween.interpolate_property(
    Pivot,
    "rotation",
    Pivot.rotation,
    MathUtil.calculate_pivot_rotation(Pivot, Mouse, Origin),
    card_swivel_dampening,
    Tween.TRANS_LINEAR,
    Tween.EASE_IN_OUT
)
```

File: src/scripts/cards/Card.gd [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+
+	if self.is_connected("body_entered", self, "_on_touchdown"):
+		self.disconnect("body_entered", self, "_on_touchdown")
+	if Area.is_connected("body_entered", self, "_on_body_entered"):
+		Area.disconnect("body_entered", self, "_on_body_entered")
+
+	unequip(card_state)
+
+
+remote func unsetup(_state):
```
Comment (Legacy107) [inline src/scripts/cards/Card.gd]: Rename `unsetup` to `reset`

File: src/scripts/cards/Card.gd [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+		return
+
+	setup(body)
+
+
+
+
+remote func setup(_Player):
+	for id in range (_Player.cards_size):
+		if _Player.cards[id] == "":
```
Comment (Legacy107) [inline src/scripts/cards/Card.gd:161]: Change default value of _Player.cards[id] to null

**Review by Legacy107 — CHANGES_REQUESTED** [review]

File: src/scripts/cards/Card.gd (hunk 4 above)
Comment (LeStolz) [reply src/scripts/cards/Card.gd]: Rename "card_info" to "card_configs" and "variable" to "config"

File: src/scripts/others/Player.gd (hunk 5 above)
Comment (LeStolz) [reply src/scripts/others/Player.gd:51]: No no no, we use PoolStringArray later so we change this later k bruh?

File: src/scripts/others/Player.gd (hunk 6 above)
Comment (LeStolz) [reply src/scripts/others/Player.gd:30]: "cards_size" can be modified by the GameHandler if we so desire in the future.

File: src/utils/MathUtil.gd (hunk 10 above)
Comment (LeStolz) [reply src/utils/MathUtil.gd]: English

File: src/scripts/cards/Card.gd (hunk 16 above)
Comment (LeStolz) [reply src/scripts/cards/Card.gd:161]: No no no We dun du dat here.

File: src/scripts/cards/Card.gd (hunk 4 above)
Comment (LeStolz) [reply src/scripts/cards/Card.gd]: Rename "card_info" to "card_settings" and "variable" to "setting"

**Review by Legacy107 — APPROVED** [review]
LGTM!

**Pull request merged**